execution/evm/vm_test.go

```diff
@@ -709,12 +717,69 @@ func TestSendCall(t *testing.T) {
 	assert.NoError(t, err, "Expected insufficient gas error")
 }
 
+// Test to ensure that contracts called with STATICCALL cannot modify state
+func TestStaticCall(t *testing.T) {
+	fmt.Printf("%s ---\n", t.Name())
+	inOff, inSize := byte(0x0), byte(0x0) // no call data
+	retOff, retSize := byte(0x0), byte(0x2)
+
+	for _, illegalOp := range []OpCode{SSTORE, LOG0, LOG1, LOG2, LOG3, LOG4, CREATE, CREATE2, SELFDESTRUCT} {
+		cache := state.NewCache(newAppState())
+		ourVm := NewVM(newParams(), crypto.ZeroAddress, nil, logger)
+
+		calleeAccount, calleeAddress := makeAccountWithCode(cache, "callee",
+			MustSplice(illegalOp, PUSH1, 0x1, return1()))
+
+		callerAccount, _ := makeAccountWithCode(cache, "caller",
+			MustSplice(PUSH1, retSize, PUSH1, retOff, PUSH1, inSize, PUSH1,
+				inOff, PUSH20, calleeAddress, PUSH1, 0x1, GAS, SUB, STATICCALL, returnWord()))
+
+		_, err := runVMWaitError(cache, ourVm, callerAccount, calleeAccount, calleeAddress, callerAccount.Code(), 1000)
+
+		for _, nestedError := range ourVm.nestedCallErrors {
+			err = nil
+			if nestedError.NestedError.ErrorCode() == errors.ErrorCodeIllegalWrite {
+				err = nestedError.NestedError.ErrorCode()
+				break
+			}
+		}
+		assert.Error(t, err, errors.ErrorCodeIllegalWrite, "Expected static call violation")
```

Comment:
This will panic if the error is not present. The signature of `assert.Error` is not what you think it is. In particular it's not like `assert.Equal` with an expected/actual argument, it just takes an error argument the rest of it is `msgAndArgs... interface{}`.

Reply:
As a side-effect of this be aware that the errors you are getting here are actually not `ErrorCodeIllegalWrite` but are related to illegal solidity (see above), at least in the examples I've checked.